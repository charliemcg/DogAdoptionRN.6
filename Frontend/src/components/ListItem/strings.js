export default {
  location: "Location:",
  price: "Price:",
  refs: {
    bounce: "BOUNCE_REF"
  },
  navigation: {
    signIn: "SignIn",
    dogProfile: "DogProfile"
  }
};
