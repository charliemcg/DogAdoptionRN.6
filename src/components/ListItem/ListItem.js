import React, {Component} from 'react';
import {
  View,
  Image,
  Text,
  TouchableHighlight,
  TouchableWithoutFeedback,
} from 'react-native';
import {connect} from 'react-redux';
import styles from './styles';
import colors from '../../colors';
import LinearGradient from 'react-native-linear-gradient';
import {selectedDog, addToFavorites, removeFromFavorites} from '../../actions';
import heartImg from '../../images/heart.png';
import heartFilledImg from '../../images/heartFilled.png';
import PropTypes from 'prop-types';
import * as Animatable from 'react-native-animatable';
import {dateFormatter} from '../../utils/dateFormatter';
import strings from './strings';
import Device from 'react-native-device-detection';

class ListItem extends Component {
  getFavoriteIcon = () => {
    return this.props.favorites.includes(this.props.item)
      ? heartFilledImg
      : heartImg;
  };

  handleFavorite = () => {
    const {favorites, item} = this.props;
    if (this.props.signedIn) {
      if (favorites.includes(item)) {
        this.props.removeFromFavorites(item);
      } else {
        //bounce the heart when adding to favorites
        this.refs[strings.refs.bounce]
          .bounce(500)
          .then(this.props.addToFavorites(item));
      }
    } else {
      this.props.navigation.navigate(strings.navigation.signIn);
    }
  };

  render() {
    const imageWrapper = (
      <View style={styles.imageWrapper}>
        <Image
          source={{uri: this.props.item.key}}
          style={styles.blurredImage}
          blurRadius={20}
        />
        <Image
          source={{uri: this.props.item.key}}
          style={styles.image}
          resizeMode="contain"
        />
        <Animatable.View ref={strings.refs.bounce} style={styles.animatable}>
          {/* icon by Smash Icons */}
          <TouchableWithoutFeedback
            style={styles.favorite}
            onPress={() => {
              this.handleFavorite();
            }}>
            <Image source={this.getFavoriteIcon()} style={styles.heart} />
          </TouchableWithoutFeedback>
        </Animatable.View>
      </View>
    );

    const quickDetailsRowOne = (
      <View style={styles.quickDetailsRow}>
        <Text style={styles.location}>
          {strings.location} {this.props.item.location}
        </Text>
        <View style={styles.price}>
          <Text>{strings.price} </Text>
          <Text style={styles.coloredPrice}>{this.props.item.price}</Text>
        </View>
      </View>
    );

    const quickDetailsRowTwo = (
      <View style={styles.quickDetailsRow}>
        <Text style={styles.breed}>{this.props.item.breed}</Text>
        <Text style={styles.date}>{dateFormatter(this.props.item.date)}</Text>
      </View>
    );

    return (
      <View style={styles.parent}>
        <TouchableHighlight
          style={styles.touchable}
          onPress={() => {
            this.props.selectedDog(this.props.item);
            this.props.navigation.navigate(strings.navigation.dogProfile);
          }}
          underlayColor={colors.notQuiteWhite}>
          <View style={styles.contentWrapper}>
            {imageWrapper}
            <View style={styles.detailsWrapper}>
              {quickDetailsRowOne}
              {quickDetailsRowTwo}
              <Text style={styles.description}>
                {this.props.item.description}
              </Text>
              <LinearGradient
                colors={
                  Device.isTablet
                    ? [
                        'rgba(238, 238, 238, 0.0)',
                        'rgba(238, 238, 238, 0.8)',
                        colors.notQuiteWhite,
                      ]
                    : [
                        'rgba(238, 238, 238, 0.0)',
                        'rgba(238, 238, 238, 0.8)',
                        colors.white,
                      ]
                }
                style={styles.fade}
              />
            </View>
          </View>
        </TouchableHighlight>
      </View>
    );
  }
}

ListItem.propTypes = {
  navigation: PropTypes.shape({
    state: PropTypes.shape({
      routeName: PropTypes.string.isRequired,
      key: PropTypes.string.isRequired,
    }).isRequired,
    actions: PropTypes.object.isRequired,
  }),
};

const mapStateToProps = state => {
  return {
    favorites: state.favorites,
    signedIn: state.signedIn,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    selectedDog: dog => {
      dispatch(selectedDog(dog));
    },
    addToFavorites: dog => {
      dispatch(addToFavorites(dog));
    },
    removeFromFavorites: dog => {
      dispatch(removeFromFavorites(dog));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(ListItem);
