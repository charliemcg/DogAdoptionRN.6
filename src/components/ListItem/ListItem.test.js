import "react-native";
import React from "react";
import ListItem from "./ListItem";
import renderer from "react-test-renderer";
import store from "../../store";

test("ListItem snapShot", () => {
  const snap = renderer
    .create(<ListItem store={store} item={jest.fn()} />)
    .toJSON();

  expect(snap).toMatchSnapshot();
});
