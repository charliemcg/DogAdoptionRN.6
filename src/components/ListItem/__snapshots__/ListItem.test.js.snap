// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListItem snapShot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#eee",
      "height": 365.625,
      "width": 750,
    }
  }
>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "height": "100%",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "height": "100%",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "height": 365.625,
            "marginBottom": "2%",
            "marginLeft": "2%",
            "marginTop": "2%",
            "width": 487.5,
          }
        }
      >
        <Image
          blurRadius={20}
          source={
            Object {
              "uri": undefined,
            }
          }
          style={
            Object {
              "flex": 1,
            }
          }
        />
        <Image
          resizeMode="contain"
          source={
            Object {
              "uri": undefined,
            }
          }
          style={
            Object {
              "flex": 1,
              "height": "100%",
              "position": "absolute",
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "height": 37.5,
              "position": "absolute",
              "right": "3%",
              "top": "5%",
              "width": 37.5,
            }
          }
        >
          <Image
            accessible={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            source={
              Object {
                "testUri": "../../../src/images/heart.png",
              }
            }
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "marginTop": "2%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": "10%",
              "marginTop": "2%",
              "width": 262.5,
            }
          }
        >
          <Text
            style={
              Object {
                "flex": 1,
                "paddingLeft": "10%",
              }
            }
          >
            Location:
             
          </Text>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": "10%",
                "textAlign": "right",
              }
            }
          >
            <Text>
              Price:
               
            </Text>
            <Text
              style={
                Object {
                  "color": "#c8a415",
                  "fontWeight": "bold",
                }
              }
            />
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "height": "10%",
              "marginTop": "2%",
              "width": 262.5,
            }
          }
        >
          <Text
            style={
              Object {
                "paddingLeft": "10%",
              }
            }
          />
          <Text
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "justifyContent": "flex-end",
                "paddingRight": "10%",
                "textAlign": "right",
              }
            }
          >
            NaN/NaN/NaN
          </Text>
        </View>
        <Text
          style={
            Object {
              "height": 450,
              "paddingTop": "2%",
              "width": 225,
            }
          }
        />
        <BVLinearGradient
          colors={
            Array [
              15658734,
              3438210798,
              4293848814,
            ]
          }
          endPoint={
            Object {
              "x": 0.5,
              "y": 1,
            }
          }
          locations={null}
          startPoint={
            Object {
              "x": 0.5,
              "y": 0,
            }
          }
          style={
            Object {
              "height": "100%",
              "position": "absolute",
              "right": 0,
              "width": 262.5,
            }
          }
        />
      </View>
    </View>
  </View>
</View>
`;
